lib/spack/spack/util/download.py

```diff
@@ -0,0 +1,345 @@
+                self.index = (self.index + 1) % len(self.spinner)
+
+            print(
+                f"\r    {percent_or_spinner}{_format_bytes(self.current_bytes)} "
+                f"@ {_format_speed(self.current_bytes, current_time - self.start_time)}"
+                f"{maybe_newline}",
+                end="",
+                flush=True,
+                file=out,
+            )
+
+
+class DownloadMethod(enum.Enum):
+    URLLIB = enum.auto()
+    CURL = enum.auto()
+
+
+class DownloadOptions(NamedTuple):
+    """Options for downloading a file."""
+
+    method: DownloadMethod = DownloadMethod.URLLIB
+    extra_args: List[str] = []
+
+
+def create_download_options(
+    method: DownloadMethod, *, extra_args: Optional[List[str]] = None
+) -> DownloadOptions:
+    """Create a DownloadOptions object from a method and extra arguments."""
+    return DownloadOptions(method=method, extra_args=extra_args or [])
+
+
+def download_file(
+    url: str,
+    *,
+    destination: str,
+    timeout: int = 0,
+    chunk_size: int = 65536,
+    options: DownloadOptions = create_download_options(DownloadMethod.URLLIB),
+) -> str:
+    """Downloads a file from the specified URL and saves it to the given path.
+
+    Args:
+        url: the URL from which the file should be downloaded.
+        destination: the local file path where the downloaded file will be saved.
+        timeout: timeout in seconds for the download.
+        chunk_size: size of chunks to read and write during the download.
+        options: options for the download of the file, including the method.
+            The default is to use urllib.
+
+    Returns:
+        The effective URL of the request.
+    """
+    if options.method == DownloadMethod.URLLIB:
+        url_reader = urllib_stream(url=url, timeout=timeout)
+    elif options.method == DownloadMethod.CURL:
+        url_reader = curl_stream(url=url, timeout=timeout, config_args=options.extra_args)
+
+    partial_file = destination + ".part"
+    with url_reader as s, open(partial_file, "wb") as f:
+        tty.msg(f"Fetching {url}")
+        _check_headers(
+            url=s.url, effective_url=s.geturl(), headers=s.headers, destination=destination
+        )
+        progress = FetchProgress.from_headers(s.headers, enabled=sys.stdout.isatty())
+        while True:
+            chunk = s.read(chunk_size)
+            if not chunk:
+                break
+            f.write(chunk)
+            progress.advance(len(chunk))
+        progress.print(final=True)
+    fs.rename(partial_file, destination)
+    return s.geturl()
+
+
+def _check_headers(
+    *, url: str, effective_url: str = "", headers: http.client.HTTPMessage, destination: str = ""
+) -> None:
+    # Check if we somehow got an HTML file rather than the archive we
+    # asked for.  We only look at the last content type, to handle
+    # redirects properly.
+    content_types = headers.get("Content-Type")
+    if content_types and "text/html" in content_types:
```

Comment:
shouldn't be plural, that's probably a remnant of the previous implementation.

```suggestion
    if "text/html" in headers.get("Content-Type", ""):
```